### PR TITLE
Feature/xnat first api

### PR DESCRIPTION
**Pull request by bertrandyvernault**
First api call, to get informations on the project

Scans and Assessors

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -59,8 +59,44 @@
 				</exclusion>
 			</exclusions>
 		</dependency>
+
+		<!-- https://mvnrepository.com/artifact/com.github.kittinunf.fuel/fuel -->
+		<dependency>
+			<groupId>com.github.kittinunf.fuel</groupId>
+			<artifactId>fuel</artifactId>
+			<version>2.2.1</version>
```
Comment (byvernault) [inline pom.xml]: Les versions sont a definir dans les properties et non ici

File: pom.xml [hunk 2]
```diff
@@ -59,8 +59,44 @@
...
+			<version>20190722</version>
+		</dependency>
+
+
 	</dependencies>
 
+	<repositories>
+		<repository>
+			<id>jcenter</id>
+			<url>https://jcenter.bintray.com/</url>
```
Comment (byvernault) [inline pom.xml]: ca sert a quoi ca ?

File: src/main/kotlin/fr/yvernault/rexa/services/scan/ScanService.kt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+                .authentication()
+                .basic(user, password)
+                .responseString() { request, response, result ->
+                    when (result) {
+                        is Result.Failure -> {
+                            val ex = result.getException()
+                            println(ex)
+                        }
+                        is Result.Success -> {
+                            val data = result.get()
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/services/scan/ScanService.kt]: I think you can simplify this by using an object ResultSet and Result and with jackson to deserialize directly your result.

https://www.baeldung.com/kotlin-fuel

File: pom.xml (hunk 2 above)
Comment (bertrandyvernault) [reply pom.xml]: I have this error :
Could not find artifact com.github.kittinunf.fuel:fuel:pom:2.2.1 in central (https://repo.maven.apache.org/maven2)

So need this repository to get it

File: src/main/kotlin/fr/yvernault/rexa/DTO/XnatInfoDTO.kt [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import fr.yvernault.rexa.model.Assessor
+import fr.yvernault.rexa.model.Scan
+
+data class  XnatInfoDTO(
+        var numberOfScan: Int,
+        var numberOfSession: Int,
+        var numberOfSubject: Int,
+        var scans: List<Scan>,
+        var assessors: List<Assessor>?
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/DTO/XnatInfoDTO.kt]: pas de ?.
On donne une liste vide si on n'a pas de donnees

File: src/main/kotlin/fr/yvernault/rexa/controller/FirstApiController.kt [hunk 5]
```diff
@@ -1,12 +1,29 @@
...
+
+        val scans = xnatService.getScansForProject(id)
+        val assessors = xnatService.getAssessorsForProject(id)
+
+        return XnatInfoDTO(
+                scans.size,
+                scans.groupBy { it.sessionLabel }.size,
+                scans.groupBy { it.subjectLabel }.size,
+                scans,
+                assessors )
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/controller/FirstApiController.kt]: espace en trop

File: src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package fr.yvernault.rexa.model
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.annotation.JsonProperty
+import fr.yvernault.rexa.XnatGlossary
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+data class Assessor(
+        @JsonProperty("ID") val id: String,
+        @JsonProperty("insert_date") val insertDate: String?,
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt]: I think it should be insertedDate

File: src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        @JsonProperty("URI") val uri: String,
+        @JsonProperty("session_ID") val sessionId: String,
+        @JsonProperty(XnatGlossary.defaultAssessorTypeLabel+ "/" + XnatGlossary.assessorStatusLabel) val procStatus: String,
+        @JsonProperty(XnatGlossary.defaultAssessorTypeLabel + "/" + XnatGlossary.assessorTypeLabel) val type: String,
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt]: I think we should renamed as procType

File: src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.annotation.JsonProperty
+import fr.yvernault.rexa.XnatGlossary
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+data class Assessor(
+        @JsonProperty("ID") val id: String,
+        @JsonProperty("insert_date") val insertDate: String?,
+        @JsonProperty("URI") val uri: String,
+        @JsonProperty("session_ID") val sessionId: String,
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt]: I will put it first just like scan

File: src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package fr.yvernault.rexa.model
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.annotation.JsonProperty
+import fr.yvernault.rexa.XnatGlossary
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+data class Assessor(
+        @JsonProperty("ID") val id: String,
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/Assessor.kt]: You should try to get as well the information on the subjectId and label and session label.

File: src/main/kotlin/fr/yvernault/rexa/model/Scan.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty
+import fr.yvernault.rexa.XnatGlossary
+
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+data class Scan(
+        @JsonProperty("session_ID") var sessionId: String,
+        @JsonProperty("ID") var id: String,
+        @JsonProperty("subject_label") var subjectLabel: String?,
+        @JsonProperty("URI") var uri: String?,
+        @JsonProperty(XnatGlossary.defaultScanTypeLabel+ "/subject_id") var subjectId: String?,
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/Scan.kt]: Warning with ?, you should not put any here.

File: src/main/kotlin/fr/yvernault/rexa/model/XnatProperties.kt [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package fr.yvernault.rexa.model
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+import org.springframework.boot.context.properties.ConstructorBinding
+
+@ConstructorBinding
+@ConfigurationProperties("xnat")
+data class XnatProperties(var url: String, val user: String,  val password: String) {
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/model/XnatProperties.kt]: should we encrypt the password here or something ?

File: src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    init {
+        FuelManager.instance.basePath = properties.url
+    }
+
+    fun getScansForProject(projectId: String): List<Scan> {
+
+        val user = properties.user
+        val password = properties.password
+
+        var sessionList: List<Scan> = emptyList()
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt]: shouldn't be named scanList or even scans?

File: src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        val httpAsync = "${XnatGlossary.urlArchiveLabel}?project=$projectId${XnatGlossary.scanUrlLabel}"
+                .httpGet()
+                .header(Headers.ACCEPT to MediaType.APPLICATION_JSON)
+                .authentication()
+                .basic(user, password)
+                .responseString() { request, response, result ->
+                    when (result) {
+                        is Result.Failure -> {
+                            val ex = result.getException()
+                            logger.error(ex.message)
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt]: We should throw an exception here if it's fail.

File: src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                            val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                            val dataJson = jsonFormat(data)
+                            assessorList = mapper.readValue(dataJson.toString())
+                        }
+                    }
+                }
+
+        httpAsync.join()
+
+        val httpAsyncProc = "${XnatGlossary.urlArchiveLabel}?project=$projectId${XnatGlossary.assessorUrlLabel}"
```
Comment (byvernault) [inline src/main/kotlin/fr/yvernault/rexa/service/XnatService.kt]: Comment the one for FreeSurfer, we will do it an other time.

File: src/main/kotlin/fr/yvernault/rexa/services/scan/ScanService.kt (hunk 3 above)
Comment (byvernault) [reply src/main/kotlin/fr/yvernault/rexa/services/scan/ScanService.kt]: lets keep it like that for now

**Pull request merged**